src/sql/src/plan/query.rs

```diff
@@ -2692,7 +2748,43 @@ impl<'a> QueryContext<'a> {
                 .chain(std::iter::once(relation_type))
                 .cloned()
                 .collect(),
+            ctes,
+        }
+    }
+
+    /// Resolves `name` to a table expr, i.e. getting a catalog table or
+    /// inlining a CTE.
+    pub fn resolve_table_expression(
+        &self,
+        name: ObjectName,
+    ) -> Result<(PartialName, RelationExpr, RelationDesc), PlanError> {
+        // Check if unqualified name refers to a CTE.
+        if name.0.len() == 1 {
+            let norm_name = normalize::ident(name.0[0].clone());
+            if let Some(cte) = self.ctes.get(&norm_name) {
+                let depth_adj = self.outer_scope.depth() - cte.original_scope_depth;
+                let mut val = cte.val.clone();
+                // Adjust column references to the correct depth; `depth_adj` >
+                // 0 iff you've entered entered subqueries.
+                val.visit_columns(depth_adj, &mut |depth, col| col.level += depth);
+
+                // Inline `val` where its name was referenced. In an ideal
+                // world, multiple instances of this expression would be
+                // de-duplicated.
+                return Ok((name.into(), val, cte.val_desc.clone()));
```

Comment:
Re comment above about normalization: what we do elsewhere (e.g. in `plan_table_alias`) is just build a `PartialName` up manually:

```rust
let name = PartialName {
    database: None,
    schema: None,
    item: norm_name,
}
```